metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/rest/client/dereference/DereferencerImpl.java

```diff
@@ -42,326 +36,415 @@
+            return new DereferencedEntities(Collections.emptyMap(), reports, classType);
+        }
+
+        // Perform the dereferencing.
+        EnrichmentResultList result;
+        String resourceId = referenceTerm.getReference().toString();
+        try {
+            LOGGER.debug("== Processing {}", resourceId);
+            result = retryableExternalRequestForNetworkExceptions(
+                    () -> dereferenceClient.dereference(resourceId));
+            DereferenceResultStatus resultStatus = Optional.ofNullable(result)
+                    .map(EnrichmentResultList::getEnrichmentBaseResultWrapperList)
+                    .orElseGet(Collections::emptyList).stream()
+                    .map(EnrichmentResultBaseWrapper::getDereferenceStatus)
+                    .filter(Objects::nonNull).findFirst()
+                    .orElse(DereferenceResultStatus.UNKNOWN_ENTITY);
+
+            setDereferenceStatusInReport(resourceId, reports, resultStatus);
+        } catch (BadRequest e) {
+            // We are forgiving for these errors
+            LOGGER.warn("ResourceId {}, failed", resourceId, e);
+            reports.add(Report
+                    .buildDereferenceWarn()
+                    .withStatus(HttpStatus.BAD_REQUEST)
+                    .withValue(resourceId)
+                    .withException(e)
+                    .build());
+            result = null;
+        } catch (Exception e) {
+            DereferenceException dereferenceException = new DereferenceException(
+                    "Exception occurred while trying to perform dereferencing.", e);
+            reports.add(Report
+                    .buildDereferenceWarn()
+                    .withStatus(HttpStatus.OK)
+                    .withValue(resourceId)
+                    .withException(dereferenceException)
+                    .build());
+            result = null;
+        }
+
+        // Return the result.
+        return new DereferencedEntities(Map.of(referenceTerm, Optional.ofNullable(result).map(EnrichmentResultList::getEnrichmentBaseResultWrapperList)
+                .orElseGet(Collections::emptyList).stream()
+                .map(EnrichmentResultBaseWrapper::getEnrichmentBaseList).filter(Objects::nonNull)
+                .flatMap(List::stream).collect(Collectors.toList())), reports, classType);
+    }
+
+    private Set<ReferenceTerm> setUpReferenceTermSet (Set<String> resourcesIds, HashSet<Report> reports){
+        return resourcesIds.stream()
+                .map(id -> checkIfUrlIsValid(reports, id))
+                .filter(Objects::nonNull)
+                .map(checkedUrl -> validateIfUrlToDereferenceExists(reports, checkedUrl))
+                .filter(Objects::nonNull)
+                .map(validateUrl -> new ReferenceTermImpl(validateUrl, new HashSet<>()))
+                .collect(Collectors.toSet());
+    }
+
+    private void updateDereferencedEntitiesMap(HashMap<Class<? extends AboutType>, DereferencedEntities> mapToUpdate,
```

Comment:
Is all this copying required?  
I think when something was not dereferenced as "Own by ID", then when it comes here(after being dereferenced with "uri" or "external") it can be simply added to the corresponding list, or am I missing something?

Reply:
When I first wrote this method, I was trying to avoid replacing values in the map with `put()` so I made all of this copying.
I suppose that after some refactoring this is not needed anymore and forget to change. I'll do that